src/porting.cpp

```diff
@@ -903,4 +908,34 @@ double perf_freq = get_perf_freq();
 
 #endif
 
+#if HAVE_MALLOC_TRIM
+
+/*
+ * On Linux/glibc we found that after deallocating bigger chunks of data (esp. MapBlocks)
+ * the memory would not be given back to the OS and would stay at peak usage.
+ * This appears to be a combination of unfortunate allocation order/fragmentation
+ * and the fact that glibc does not call madvise(MADV_DONTNEED) on its own.
+ * Some other allocators were also affected, jemalloc and musl libc were not.
+ * read more: <https://forum.minetest.net/viewtopic.php?t=30509>
+ *
+ * As a workaround we track freed memory coarsely and call malloc_trim() once a
+ * certain amount is reached.
+ */
+
+static std::atomic<size_t> memory_freed;
+
+constexpr size_t MEMORY_TRIM_THRESHOLD = 128 * 1024 * 1024;
+
+void TrackFreedMemory(size_t amount)
+{
+	size_t sum = memory_freed.fetch_add(amount, std::memory_order_relaxed);
+	if (sum > MEMORY_TRIM_THRESHOLD) {
+		// Leave some headroom for future allocations
+		malloc_trim(1 * 1024 * 1024);
+		memory_freed.store(0);
+	}
```

Comment:
I don't understand the first suggestion but it makes even less sense to me combined with another `.load()`: https://godbolt.org/z/6vT7MYfj1

Reply:
Fech-add (i.e. `lock xadd` in x86-64) returns the value before the operation, and does this atomically (nothing can come in-between). A separate load and atomic add (i.e. `lock add` and `mov`) can also be implemented via a fetch-add (by the compiler or the cpu). It's the same but less strict. And we don't require the fetch-add to be atomic here.
Idk if this makes much if any performance difference in practice.

Reply:
I read the description of `fetch_add` several times because it was written so weirdly and didn't realize that it returned the old value.
In that case I'll just use the separate load & add.